### PR TITLE
Revert "Revert "Fix unexpected behavior with `FORMAT` and `SETTINGS` parsing""

### LINKED ISSUE #39712: FORMAT and SETTINGS parsing leads to errors
Query parsing related to `FORMAT` and `SETTINGS` clauses is inconsistent.

`select 1 settings log_comment='test' Format {{format}}` works as intended
`(select 1) settings log_comment='test' Format {{format}}` and `(select 1) union (select 2) settings log_comment='test' Format {{format}}` give a parsing error:

```
Syntax error: failed at position 50 ('Format'):

(select 1)  settings log_comment='prova valerio' Format JSONCompact

Expected end of query
```

Queries without the `FORMAT` clause work as intended.

**Does it reproduce on recent release?**
22.7.1.2484 is affected

**How to reproduce**
Connect to clickhouse-client and run the queries

**Expected behavior**

Same behaviour between different versions

> If applicable, add screenshots to help explain your problem.

image

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Reverts ClickHouse/ClickHouse#68608

### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to write `SETTINGS` before `FORMAT` in a chain of queries with `UNION` when subqueries are inside parentheses. This closes #39712. Change the behavior when a query has the SETTINGS clause specified twice in a sequence. The closest SETTINGS clause will have a preference for the corresponding subquery. In the previous versions, the outermost SETTINGS clause could take a preference over the inner one.

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [x]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 314f2b30d116f0995c14a7a6747c407ec7c81df3 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68614/314f2b30d116f0995c14a7a6747c407ec7c81df3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d80b37001c10370c1e91a1fb9d2560fbc7ea1e8d/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/aa0e4beae6894abd49922233bbde5187b398180d/fast_test.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/aa0e4beae6894abd49922233bbde5187b398180d/unit_tests__release_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Client/ClientBase.cpp [hunk 1]
```diff
@@ -1892,48 +1893,19 @@ void ClientBase::processParsedSingleQuery(const String & full_query, const Strin
...
-                {
-                    apply_query_settings(*last_select->settings());
-                }
-            }
-        }
-        else if (const auto * query_with_output = parsed_query->as<ASTQueryWithOutput>(); query_with_output && query_with_output->settings_ast)
-            apply_query_settings(*query_with_output->settings_ast);
-        else if (insert && insert->settings_ast)
-            apply_query_settings(*insert->settings_ast);
+            client_context->setSettings(old_settings);
```
Comment (vdimir) [inline src/Client/ClientBase.cpp:1900]: After some debugging, I suspect `00653_verification_monotonic_data_load` timeouts under tsan caused by this line, client hung here for some reason. However, no explanation why it reproduced only at one test and didn't reproduce before, even though we had `setSettings` under `if (old_settings)`

Comment (vdimir) [reply src/Client/ClientBase.cpp:1900]: Also, the issue might be related to grep -q, such as early termination of the client leading to that

Comment (vdimir) [reply src/Client/ClientBase.cpp:1900]: There’s a data race that this PR reveals
https://github.com/ClickHouse/ClickHouse/issues/68626#issuecomment-2305120194

We can merge this PR after fixing it by https://github.com/ClickHouse/ClickHouse/pull/68791

Comment (thevar1able) [reply src/Client/ClientBase.cpp:1900]: https://github.com/ClickHouse/ClickHouse/pull/69186 is merged, are we missing anything else?

**Comment by thevar1able** [comment]
FastTest is still :red_circle:.

**Comment by vdimir** [comment]
> Stateless tests (tsan, s3 storage) [2/3] — fail: 1, passed: 2321, skipped: 68 [Details](https://s3.amazonaws.com/clickhouse-test-reports/68614/1ee3cea4001ee06f8e90ba1f742ef2501837d256/stateless_tests__tsan__s3_storage__[2_3].html)

That's werid, like data race still reproduces... I'm going to take a look...

**Comment by alexey-milovidov** [comment]
@vdimir, it does not look like the case.

**Comment by Algunenano** [comment]
Seems like this might be causing `00653_verification_monotonic_data_load` to fail again

**Comment by tavplubix** [comment]
Yes, it breaks the test just like the [original PR](https://github.com/ClickHouse/ClickHouse/pull/60197): [CI DB](https://play.clickhouse.com/play?user=play#CldJVEgKICAgICcwMDY1M192ZXJpZmljYXRpb25fbW9ub3RvbmljX2RhdGFfbG9hZCcgQVMgbmFtZV9zdWJzdHIsCiAgICA5MCBBUyBpbnRlcnZhbF9kYXlzLAogICAgKCdTdGF0ZWxlc3MgdGVzdHMgKGFzYW4pJywgJ1N0YXRlbGVzcyB0ZXN0cyAoYWRkcmVzcyknLCAnU3RhdGVsZXNzIHRlc3RzIChhZGRyZXNzLCBhY3Rpb25zKScsICdJbnRlZ3JhdGlvbiB0ZXN0cyAoYXNhbikgWzEvM10nLCAnU3RhdGVsZXNzIHRlc3RzICh0c2FuKSBbMS8zXScpIEFTIGJhY2twb3J0X2FuZF9yZWxlYXNlX3NwZWNpZmljX2NoZWNrcwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkLAogICAgY291bnQoKSwKICAgIGdyb3VwVW5pcUFycmF5KHB1bGxfcmVxdWVzdF9udW1iZXIpIEFTIHBycywKICAgIGFueShyZXBvcnRfdXJsKQpGUk9NIGNoZWNrcwpXSEVSRSAoKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZSkgQU5EIChwdWxsX3JlcXVlc3RfbnVtYmVyIE5PVCBJTiAoCiAgICBTRUxFQ1QgcHVsbF9yZXF1ZXN0X251bWJlciBBUyBwcm4KICAgIEZST00gY2hlY2tzCiAgICBXSEVSRSAocHJuICE9IDApIEFORCAoKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZSkgQU5EIChjaGVja19uYW1lIElOIChiYWNrcG9ydF9hbmRfcmVsZWFzZV9zcGVjaWZpY19jaGVja3MpKQopKSBBTkQgKHBvc2l0aW9uKHRlc3RfbmFtZSwgbmFtZV9zdWJzdHIpID4gMCkgQU5EICh0ZXN0X3N0YXR1cyBJTiAoJ0ZBSUwnLCAnRVJST1InLCAnRkxBS1knKSkKR1JPVVAgQlkgZApPUkRFUiBCWSBkIERFU0MK)

**Pull request merged**